Validate command line flags

vanosg opened the pull request:
Found by: Geo
Patch by: Geo
Fixes: #359 

One-line summary:
Improve command line parsing

Additional description (if needed):


Test cases demonstrating functionality (if applicable):
```
$ ./eggdrop -v
Eggdrop v1.8.1+alpha (C) 1997 Robey Pointer (C) 1999-2017 Eggheads
Configure flags: none
Compiled with: IPv6, TLS, handlen=32
```

```
$ ./eggdrop -vj
./eggdrop: invalid option -- 'j'

Eggdrop v1.8.1+alpha (C) 1997 Robey Pointer (C) 1999-2017 Eggheads

Usage: eggdrop [options] [config-file]

Options:
  -n   Don't background; send all log entries to console.
  -nc  Don't background; display channel stats every 10 seconds.
  -nt  Don't background; use terminal to simulate DCC chat.
  -m   Create userfile.
  -h   Show this help.
  -v   Show version info, then quit.
```

etc...

vanosg committed:
Remove rogue -h entry from english help file

vanosg committed:
Improve command line flag parsing
Use getopt() to parse command line
Disallow invalid flags from being given (previously ignored) on command line

Cizzle commented on src/main.c:
```diff
@@ -52,6 +52,7 @@
 #include <signal.h>
 #include <netdb.h>
 #include <setjmp.h>
+#include <getopt.h>
```
The "getopt" fuction is located in the unistd.h header making this getopt.h include unnecessary (though it also contains the prototype).

Cizzle commented on src/main.c:
```diff
@@ -496,56 +497,104 @@ void eggAssert(const char *file, int line, const char *module)
+   | QUIT| BAD FLAG| h| n| c| t| m| v|
+   |  128|       64|32|16| 8| 4| 2| 1|
+*/
+  int cliflags = 0;
```
An `unsigned char` should be enough?

Cizzle commented on src/main.c:
```diff
@@ -496,56 +497,104 @@ void eggAssert(const char *file, int line, const char *module)
+        if (cliflags & 16) {
+	  break;
+ 	}
+	cliflags = cliflags | 16;
```
I'd use
```
if (~(cliflags & 16)) {
  cliflags |= 16
  do_other_stuff;
}
break;
```
for this and all other cases.

Cizzle commented on src/main.c:
```diff
@@ -496,56 +497,104 @@ void eggAssert(const char *file, int line, const char *module)
+    show_ver();
+    exit(0);
+  } else {
     strncpyz(configfile, s, sizeof configfile);
```
Could use `argv[optind]` (including a check of `optind < argc`) instead of `s` and then do away with the `do_arg()` argument and the loop in main where `do_arg()` is called from?

vanosg committed:
code cleanup

vanosg committed:
Remove old argv loops

vanosg committed:
Add warning for multiple values of config file

vanosg committed:
Refactor cliflag bitmask break logic

vanosg committed:
Remove lang file dependency for -h

vanosg committed:
Check args before initializing language
Prevents -h/-v from showing unneeded error message if lang files are not in default location, as config file has not yet been parsed to update with alternate location

vanosg committed:
Hard code CLI help, as language files aren't yet loaded for use

vanosg committed:
Re-Refactor bitmask logic- now with MOAR logic

vanosg committed:
Fix multi-config value error

vanosg committed:
Fix multi-config value error for realz

vanosg committed:
Fix spacing/comments

Cizzle commented on src/main.c:
```diff
@@ -496,56 +496,99 @@ void eggAssert(const char *file, int line, const char *module)
+/* Bitmask structure to hold cli flags
+   | QUIT| BAD FLAG| h| n| c| t| m| v|
+   |  128|       64|32|16| 8| 4| 2| 1|
+*/
```
Might be better to put this in an enum or at least #define's?

Cizzle commented on src/main.c:
```diff
@@ -496,56 +496,99 @@ void eggAssert(const char *file, int line, const char *module)
+      printf("WARNING: More than one config file value detected\n");
+      printf("         Using %s as config file\n", argv[optind]);
+    }
+    strncpyz(configfile, argv[optind], sizeof configfile);
```
What will configfile get when I run `./eggdrop -m`?

vanosg replied:
Should still work based on ```if (argc > (optind + 1)) {``` ... ie, if the total number of args given is than 2 more than the number of flags

```
$ ./eggdrop -m patchtest.conf 

Eggdrop v1.8.1+alpha (C) 1997 Robey Pointer (C) 1999-2017 Eggheads
[09:44:15] --- Loading eggdrop v1.8.1+alpha (Tue Apr 25 2017)
[09:44:15] Listening at telnet port 3183 (all).
[13:44:15] Module loaded: dns             
[13:44:15] Module loaded: channels        
[13:44:15] Module loaded: server          
[13:44:15] Module loaded: ctcp            
[13:44:15] Module loaded: irc             
[13:44:15] Module loaded: transfer         (with lang support)
[13:44:15] Module loaded: share           
[13:44:15] Module loaded: notes            (with lang support)
[13:44:15] Module loaded: console          (with lang support)
[13:44:15] Module loaded: blowfish        
[13:44:15] Writing channel file...
[13:44:15] Userfile loaded, unpacking...
USERFILE ALREADY EXISTS (drop the '-m')
[13:44:15] === scooby: 0 channels, 130 users.
Launched into the background  (pid: 64929)
```

vanosg replied:
Damnit...  :stuck_out_tongue: 
```
$ ./eggdrop -m
[09:49:52] * Last context: main.c/1036 []
[09:49:52] * Please REPORT this BUG!
[09:49:52] * Check doc/BUG-REPORT on how to do so.
Segmentation fault (core dumped)
```

vanosg committed:
Fix using default eggdrop.conf if no conf file given

vanosg committed:
Final(?) refactor of multi-config values given

vanosg commented:
build

vanosg commented:
build

Cizzle commented:
LGTM.

vanosg committed:
Merge branch 'develop' into validatecliflags

vanosg committed:
Update patchlevel
